src/core/mesh/qgsmeshdataprovider.h

```diff
@@ -227,6 +227,93 @@ class CORE_EXPORT QgsMeshDataBlock
     DataType mType;
 };
 
+/**
+ * \ingroup core
+ *
+ * QgsMesh3dDataBlock is a block of 3d stacked mesh data related N
+ * faces defined on base mesh frame.
+ *
+ * data are implicitly shared, so the class can be quickly copied
+ * std::numeric_limits<double>::quiet_NaN() represents NODATA value
+ *
+ * \since QGIS 3.12
+ */
+class CORE_EXPORT QgsMesh3dDataBlock
+{
+  public:
+    //! Type of data stored in the block
+    enum DataType
+    {
+      ActiveFlagInteger, //!< Integer boolean flag whether volume is active
+      ScalarDouble, //!< Scalar double values
+      VectorDouble, //!< Vector double pairs (x1, y1, x2, y2, ... )
+      VerticalLevelsCount, //! Integer number of vertical levels above face
+      VerticalLevels, //! Double - size + 1
+      FaceToVolumeIndex //! Integer
+    };
+
+    //! Constructs an invalid block
+    QgsMesh3dDataBlock();
+
+    //! Dtor
+    ~QgsMesh3dDataBlock();
+
+    //! Constructs a new block for count faces
+    QgsMesh3dDataBlock( int count, int maximumVerticalLevels, bool isVector );
+
+    //! Sets block as valid
+    void setIsValid();
+
+    //! Whether the block is valid
+    bool isValid() const;
+
+    //! Whether we store vector values
+    bool isVector() const;
+
+    //! Number of items stored in the block
+    int count() const;
+
+    //! Index of the first volume stored in the buffer
+    int firstVolumeIndex() const;
+
+    //! Index of the last volume stored in the buffer
+    int lastVolumeIndex() const;
+
+    /**
+     * Returns internal buffer to the array
+     *
+     * The buffer is already allocated with size:
+     * count() * sizeof(int) for ActiveFlagInteger
+     * count() * sizeof(double) for ScalarDouble
+     * count() * 2 * sizeof(double) for VectorDouble
+     * count() * sizeof(int) for VerticalLevelsCount
+     * count() * sizeof(int) for FaceToVolumeIndex
+     * count() * mMaximumVerticalLevels * sizeof(double) for VerticalLevels
+     *
+     * Primary usage of the function is to write/populate
+     * data to the block by data provider.
+     */
+    void *buffer( DataType type ) SIP_SKIP;
```

Comment:
As discussed with Peter, this would be good to split into multiple functions - but it seems it would be best to leave this for a follow up PR which will also make 'active' flag optional.